drivers/scheduler/k8s/k8s.go

```diff
@@ -5134,6 +5144,73 @@ func (k *K8s) createAdmissionRegistrationObjects(
 	return nil, nil
 }
 
+// createVirtualMachineObjects creates the kubevirt VirtualMachines using kubectl apply
+func (k *K8s) createVirtualMachineObjects(
+	spec interface{},
+	ns *corev1.Namespace,
+	app *spec.AppSpec,
+) (interface{}, error) {
+	if obj, ok := spec.(*kubevirtv1.VirtualMachine); ok {
+		// Validating to make sure the desired images are imported by CDI and volumes are ready to be used by Kubevirt VM
+		virtualMachineVolumes := obj.Spec.Template.Spec.Volumes
+		if len(virtualMachineVolumes) > 0 {
+			for _, v := range virtualMachineVolumes {
+				// Validating Volume Source of type PersistentVolumeClaim is ready to be used by Kubevirt VM
+				if v.VolumeSource.PersistentVolumeClaim != nil {
+					pvcName := v.VolumeSource.PersistentVolumeClaim.ClaimName
+					t := func() (interface{}, bool, error) {
+						events, err := k8sCore.ListEvents(ns.Name, metav1.ListOptions{
+							FieldSelector: fmt.Sprintf("involvedObject.kind=PersistentVolumeClaim,involvedObject.name=%s", pvcName),
+						})
+						if err != nil {
+							return "", false, err
+						}
+						log.Infof("Events for pvc [%s] in namespace [%s] for virtual machine [%s] \n%v\n", pvcName, ns.Name, obj.Name, events)
+						for _, event := range events.Items {
+							if strings.Contains(event.Message, cdiImportSuccessEvent) {
+								pvc, err := k8sCore.GetPersistentVolumeClaim(pvcName, ns.Name)
+								if err != nil {
+									return "", false, err
+								}
+								log.Infof("cdi.kubevirt.io/storage.condition.running.message status - [%s]", pvc.Annotations[cdiPvcRunningMessageAnnotationKey])
+								return "", false, nil
+							}
+						}
+						return "", true, fmt.Errorf("waiting for import to be completed for pvc [%s] in namespace [%s] for virtual machine [%s]", pvcName, ns.Name, obj.Name)
+					}
+					_, err := task.DoRetryWithTimeout(t, 5*time.Minute, 30*time.Second)
+					if err != nil {
+						return nil, err
+					}
+				}
+				// TODO: For other Volume Source types like Data Volumes, validation logic should come here
+			}
+		}
+		// Create VirtualMachine Spec
+		if obj.Namespace != "kube-system" {
+			obj.Namespace = ns.Name
+		}
+		vm, err := k8sKubevirt.CreateVirtualMachine(obj)
+		if k8serrors.IsAlreadyExists(err) {
+			if vm, err = k8sKubevirt.GetVirtualMachine(obj.Name, obj.Namespace); err == nil {
+				log.Infof("[%v] Found existing VirtualMachine: %v", app.Key, obj.Name)
+				return vm, nil
+			}
```

Comment:
How about else return nil, err?

Reply:
We have to return the VM object if there is already one existing

Reply:
err from k8sKubevirt.GetVirtualMachine(obj.Name, obj.Namespace)